Add locking functionality

SkydiveMarius opened the pull request:
The goal of this patch is to add locking functionality as discussed in issue #37.
Only standalone / replicated server setups are supported, as locking in active/active clusters without race conditions or massive performance impact is not possible in my eyes.

The patch follows the approach of adding an additional time limited Redis key/value pair
- Key: <php_session_key>_LOCK
- Value: Process unique SHA1 hash (PID + Hostname)
And using LUA Script for releasing as described  in the official [Redis documentation.](https://redis.io/topics/distlock#correct-implementation-with-a-single-instance).

The current implementation makes use of four new INI parameters with the following default values:
```ini
# Enables/Disables locking at all
redis.session.locking_enabled=0

# Maximum retries of locking attempts until ignoring the current lock
redis.session.lock_retries=10

# Sleep time in microseconds between locking attempts 
redis.session.lock_wait_time=2000

# TTL of lock key/value pairs, if not set or Zero, then max_execution_time is used 
redis.session.lock_expire=
```

Thanks to @KiNgMaR for lots of support and concept creation.

As ny main competence is in PHP not in C, there is much space for improvement, so please code review has hard as you want :)

michael-grunder commented:
Thanks for the pull request!  I think a lot of people who use phpredis for sessions will appreciate it greatly.

I'll pull this down locally and go through it this evening.

yatsukhnenko commented:
@michael-grunder, unfortunately I can't test it right now but I can do some code review

yatsukhnenko commented on redis_session.c:
```diff
@@ -183,6 +186,130 @@ redis_pool_get_sock(redis_pool *pool, const char *key TSRMLS_DC) {
+    PHP_SHA1Init(&sha_context);
+    PHP_SHA1Update(&sha_context, lock_secret.c, lock_secret.len);
+    PHP_SHA1Final(sha_digest, &sha_context);
+    make_sha1_digest(lock_secret_hash, sha_digest);
```
Is it necessary to use hashing here? Why don't you use hostname + pid string?

SkydiveMarius replied:
You are right, its not necessary anymore. In the beginning also a random number was part of a secret, which was the reason for hashing.
Will be removed.

yatsukhnenko commented on redis_session.c:
```diff
@@ -183,6 +186,130 @@ redis_pool_get_sock(redis_pool *pool, const char *key TSRMLS_DC) {
+        smart_string_appendl(&lock_key, "_LOCK", strlen("_LOCK"));
+        smart_string_0(&lock_key);
+        strncpy(lock_status->lock_key, lock_key.c, 256);
+
```
I guess you need to free memory that was allocated for `lock_key`

SkydiveMarius replied:
You are right, I also spotted  another memory leaking smart_string. Will be fixed.

yatsukhnenko commented on redis_session.c:
```diff
@@ -306,11 +437,18 @@ PS_OPEN_FUNC(redis)
 PS_CLOSE_FUNC(redis)
 {
     redis_pool *pool = PS_GET_MOD_DATA();
+    redis_pool_member *rpm = redis_pool_get_sock(pool, pool->lock_status->session_key TSRMLS_CC);
```
You should move all code below into the `if(pool){` block

SkydiveMarius replied:
Thanks, will be done.

yatsukhnenko commented on redis_session.c:
```diff
@@ -361,9 +499,14 @@ PS_READ_FUNC(redis)
     resp = redis_session_key(rpm, key->val, key->len, &resp_len);
 #endif
-    cmd_len = REDIS_SPPRINTF(&cmd, "GET", "s", resp, resp_len);
+    pool->lock_status->session_key = (char *) malloc(resp_len);
```
Why don't you use `emalloc` here?

SkydiveMarius replied:
Thanks, was not aware of that. Will be replaced.

yatsukhnenko commented on redis_session.c:
```diff
@@ -426,7 +569,12 @@ PS_WRITE_FUNC(redis)
                              INI_INT("session.gc_maxlifetime"), val->val,
                              val->len);
 #endif
+    pool->lock_status->session_key = (char *) malloc(strlen(session));
```
Why don't you use `emalloc` here?

SkydiveMarius replied:
Will be replaced.

yatsukhnenko commented on redis_session.c:
```diff
@@ -183,6 +186,130 @@ redis_pool_get_sock(redis_pool *pool, const char *key TSRMLS_DC) {
+            REDIS_G(lock_release_lua_script_uploaded) = 0;
+            upload_lock_release_script(redis_sock);
+            redis_sock_write(redis_sock, cmd, cmd_len TSRMLS_CC);
+            response = redis_sock_read(redis_sock, &response_len TSRMLS_CC);
```
`redis_sock_read` may return `NULL`

SkydiveMarius replied:
Thank you for the hint. In this particular case the response is not used.
But in other cases, as for example LUA script upload, we do. How would you suggest should we handle errors like NULL return there (Warning, failure return, etc)?

yatsukhnenko replied:
You do `efree(response)` below and it doesn't like when argument is `NULL` :)

SkydiveMarius replied:
Thanks, I see. Will be fixed.

yatsukhnenko commented on redis_session.c:
```diff
@@ -183,6 +186,130 @@ redis_pool_get_sock(redis_pool *pool, const char *key TSRMLS_DC) {
+        strncpy(lock_status->lock_key, lock_key.c, 256);
+
+        if (lock_expire > 0) {
+            cmd_len = REDIS_SPPRINTF(&cmd, "SET", "ssssd", lock_status->lock_key, strlen(lock_status->lock_key), lock_status->lock_secret_hash, strlen(lock_status->lock_secret_hash), "NX", 2, "PX", 2, lock_expire * 1000);
```
Is the multiplication to 1K necessary here? May be `redis.session.lock_expire` should be in milliseconds?

SkydiveMarius replied:
By default max_executim_time is used as TTL, which is in seconds. Therefore I though I may be more consistent to keep redis.session.lock_expire also in seconds.
Do you think its better to have redis.session.lock_expire in milliseconds? But we still have to do the 1K multiplication for the default behavior with max_execution_time.

yatsukhnenko commented on redis_session.c:
```diff
@@ -183,6 +185,127 @@ redis_pool_get_sock(redis_pool *pool, const char *key TSRMLS_DC) {
+{
+    int locking_enabled = INI_INT("redis.session.locking_enabled");
+
+    if (!lock_status->is_locked && locking_enabled > 0) {
```
Don't you want to use
```
if (lock_status->is_locked || INI_INT()) return;
```
and move all other code outside from `if` statement?

SkydiveMarius replied:
Good idea, fixed that and also another similar spot.

1ma commented:
Glad to see session locking going into the native extension :) I have a couple of questions regarding the INI params.
```ini
# Enables/Disables locking at all
redis.session.locking_enabled=0
```
I understand this is for backwards compatibility but since lockless sessions are an actual bug, does it really make sense to allow this behavior from now on?
```ini
# Maximum retries of locking attempts until ignoring the current lock
redis.session.lock_retries=10
```
Reading the `lock_acquire` function it looks like it will try to acquire the lock at most `lock_retries` times, then it will consider that the lock was acquired even if all attempts failed (!). Instead of having a `lock_retries` param, I think it'd be more correct to just loop forever until success. Since locks are being created with a TTL they are guaranteed to disappear after a certain time has passed, even if the PHP process that created them fails to run `session_write_close()`.

yatsukhnenko commented:
> I think it'd be more correct to just loop forever until success

Forever loop is really bad idea. After `lock_retries` number of tries locking should failes.

SkydiveMarius commented:
@yatsukhnenko With fail you mean, that the process should ignore locking and and continue to force read/write session data or that the complete read/write operation should fail?

1ma commented:
But it's not really forever since the locks have a TTL, eventually the operation must succeed (*unless* the exclusive `SET` command can fail for other reasons, even if the lock for the same session does not exist, this I don't know).

At any rate it looks like [now it's not failing](https://github.com/SkydiveMarius/phpredis/blob/8c5a450f9dab518d6ccab6a026e6c925de181189/redis_session.c#L232), it just goes ahead not matter what.

yatsukhnenko commented:
@SkydiveMarius 
> the complete read/write operation should fail

SkydiveMarius commented:
I changed the behavior, that in case of exceeded lock_retries the complete read/write operation will fail.

In this context I discovered two edge case: 
- If TTL < max_execution_time, it was possible that before writing session data the lock expired (and maybe an other process acquired it already).
- On PHP < 7.0, PHP core does not verify the status of external session handlers properly. So it was possible that on failed READ (e.g. caused by exceeded lock_retries), PHP core nevertheless tries to write session data.

Therefore I implemented a lock holding check before writing session data.
Both behaviors are now covered by unit tests.

yatsukhnenko commented:
@SkydiveMarius, you don't check returned by `lock_acquire` value

yatsukhnenko commented:
Ah, no. I'm sorry. Looked at wrong commit...

yatsukhnenko commented:
Could you squash all your commits when all work will be done?

SkydiveMarius commented:
@yatsukhnenko: Alright, we will start now to test the patch in our test, pre-production and production environment during the next days/weeks.
As soon I am sure that its basically stable, I will squash all commits.

yatsukhnenko commented:
@SkydiveMarius how is your tests? Could you rebase and squash your changes?

SkydiveMarius commented:
@yatsukhnenko Unfortunately we had some delays.
Tests have been started yesterday with ~60 BETA test developers.

I will keep the PR up2date with the test results.
As soon we are sure about the stability, I will rebase and squash.

SkydiveMarius commented:
I am happy to inform you that our tests have been successful and from our point of view the patch seems stable and ready for the next RC.

**Test environment:**
PHP Version: 5.5 & 7.1
Type: PHP-FPM & PHP-CLI
Beta Testers: ~60 developers
Application servers: ~12
Active parallel sessions: ~9k

@yatsukhnenko Feature branch has been rebased and all commits have been squashed.

yatsukhnenko commented:
@SkydiveMarius travisci checks failed

SkydiveMarius commented:
@yatsukhnenko Thanks, seems there is an issue regarding ZTS.
Currently working on this...

SkydiveMarius commented:
@yatsukhnenko Issues regarding ZTS have been fixed and Travis builds are green again.

yatsukhnenko commented on redis_session.c:
```diff
@@ -361,9 +530,14 @@ PS_READ_FUNC(redis)
     resp = redis_session_key(rpm, ZSTR_VAL(key), ZSTR_LEN(key), &resp_len);
 #endif
-    cmd_len = REDIS_SPPRINTF(&cmd, "GET", "s", resp, resp_len);
+    pool->lock_status->session_key = (char *) emalloc(resp_len + 1);
```
Use `estrndup` here

SkydiveMarius replied:
Will be fixed.

yatsukhnenko commented on redis_session.c:
```diff
@@ -361,9 +530,14 @@ PS_READ_FUNC(redis)
+    cmd_len = REDIS_SPPRINTF(&cmd, "GET", "s", resp, resp_len);
     efree(resp);
+
+    lock_acquire(redis_sock, pool->lock_status TSRMLS_CC);
```
Why do you ignore result of `lock_acquire`?

SkydiveMarius replied:
Because it does not matter for the current logic.
But I think it makes sense to throw a warning.

In the current logic, a failing lock_aqauire (e.g. redis.session.lock_retries reached) will allow reads but will deny writes.
Do you think we should change this?

yatsukhnenko commented on redis_session.c:
```diff
@@ -183,6 +185,162 @@ redis_pool_get_sock(redis_pool *pool, const char *key TSRMLS_DC) {
+
+void calculate_lock_secret(redis_session_lock_status *lock_status)
+{
+    char hostname[64] = {0};
```
I think you should use `HOST_NAME_MAX` value here. 64 bytes may be not enough.

SkydiveMarius replied:
Thanks, will be fixed.

yatsukhnenko commented on redis_session.c:
```diff
@@ -183,6 +185,162 @@ redis_pool_get_sock(redis_pool *pool, const char *key TSRMLS_DC) {
+    gethostname(hostname, 64);
+
+    // Concatenating the redis lock secret
+    smart_string_appendl(&lock_status->lock_secret, hostname, strlen(hostname));
```
Why don't you use `spprintf`?

SkydiveMarius replied:
Thanks, will be fixed.

yatsukhnenko commented on redis_session.c:
```diff
@@ -183,6 +185,162 @@ redis_pool_get_sock(redis_pool *pool, const char *key TSRMLS_DC) {
+
+    cmd_len = REDIS_SPPRINTF(&cmd, "SCRIPT", "ss", "LOAD", strlen("LOAD"), release_script, strlen(release_script));
+
+    redis_sock_write(redis_sock, cmd, cmd_len TSRMLS_CC);
```
Don't ignore return value of `redis_sock_write`

SkydiveMarius replied:
Will be fixed + warning.

yatsukhnenko commented on redis_session.c:
```diff
@@ -183,6 +185,162 @@ redis_pool_get_sock(redis_pool *pool, const char *key TSRMLS_DC) {
+
+    if (response != NULL) {
+        memset(REDIS_G(lock_release_lua_script_hash), 0, 41);
+        strncpy(REDIS_G(lock_release_lua_script_hash), response, strlen(response));
```
Use `estrdup` here

SkydiveMarius replied:
Will be fixed.

yatsukhnenko commented on redis_session.c:
```diff
@@ -183,6 +185,162 @@ redis_pool_get_sock(redis_pool *pool, const char *key TSRMLS_DC) {
+        upload_lock_release_script(redis_sock TSRMLS_CC);
+        cmd_len = REDIS_SPPRINTF(&cmd, "EVALSHA", "sdss", REDIS_G(lock_release_lua_script_hash), strlen(REDIS_G(lock_release_lua_script_hash)), 1, lock_status->lock_key.c, lock_status->lock_key.len, lock_status->lock_secret.c, lock_status->lock_secret.len);
+
+        redis_sock_write(redis_sock, cmd, cmd_len TSRMLS_CC);
```
Don't ignore return value of `redis_sock_write`

SkydiveMarius replied:
Will be fixed + additional error handling to check if just LUA script was not uploaded or an other error occured.

yatsukhnenko commented on redis_session.c:
```diff
@@ -183,6 +185,162 @@ redis_pool_get_sock(redis_pool *pool, const char *key TSRMLS_DC) {
+
+    cmd_len = REDIS_SPPRINTF(&cmd, "GET", "s", lock_status->lock_key.c, lock_status->lock_key.len);
+
+    redis_sock_write(redis_sock, cmd, cmd_len TSRMLS_CC);
```
Don't ignore return value of redis_sock_write

SkydiveMarius replied:
Will be fixed. I think in addition a warning makes sense.
In case of error (response == null), is_locked is set to false anyway and write operation will be denied.

yatsukhnenko commented on redis.c:
```diff
@@ -73,6 +73,12 @@ PHP_INI_BEGIN()
     PHP_INI_ENTRY("redis.clusters.timeout", "", PHP_INI_ALL, NULL)
+
+    /* redis session */
+    PHP_INI_ENTRY("redis.session.locking_enabled", "", PHP_INI_ALL, NULL)
```
Sort ini entries alphabetically

SkydiveMarius replied:
Will be fixed.

yatsukhnenko commented on php_redis.h:
```diff
@@ -260,9 +260,17 @@ PHP_REDIS_API int redis_sock_read_multibulk_multi_reply_loop(
-#ifndef _MSC_VER
 ZEND_BEGIN_MODULE_GLOBALS(redis)
+    int lock_release_lua_script_uploaded;
+    char lock_release_lua_script_hash[41];
```
Why this is not just a pointer?

SkydiveMarius replied:
To avoid the need for later cleanup when process ends.
Do you see advantages in using a pointer?

yatsukhnenko replied:
Sounds reasonable

SkydiveMarius replied:
In the end I switched to pointer, because your suggestion with using estrdup brings much more simplicity and is only possible with pointer.

SkydiveMarius commented:
@yatsukhnenko: For better tracking, I will use again single commits for processing the code review. 
As soon as you say the review has finished I will squash again.

yatsukhnenko commented on common.h:
```diff
@@ -15,6 +15,7 @@ typedef smart_str smart_string;
 #define smart_string_appendc(dest, c) smart_str_appendc(dest, c)
 #define smart_string_append_long(dest, val) smart_str_append_long(dest, val)
 #define smart_string_appendl(dest, src, len) smart_str_appendl(dest, src, len)
+#define smart_string_free(str) smart_str_free(str)
```
We don't need this anymore :)

SkydiveMarius replied:
Nice, feels good to delete obsolete code :+1:

yatsukhnenko commented:
Looks good to me, @SkydiveMarius good job! :+1: 
@michael-grunder could you take a glance?

michael-grunder commented:
I've pulled the branch in locally and have made a few small changes having mostly to do with formatting but also a couple things to make the code more "C like" here and there.

@SkydiveMarius There is quite a lot of logic around attempting to `SCRIPT LOAD` the lock release LUA, and then if we fail the upload is attempted again (e.g. if the LUA script cache was flushed).

I think a different strategy would simplify the code on our end and might be better but wanted to run it by everyone in case there are things I'm not considering.

Instead of attempting a `SCRIPT LOAD` multiple times, what about simply computing the SHA1 hash and then doing the following on a lock release:

First attempt to run EVALSHA.  If that fails then send the LUA itself via EVAL.  This way we can't have a race condition where something like this happens:

1.  We send script load
2.  Somebody else flushes the LUA
3.  We attempt the EVALSHA (this fails)
4.  We attempt to SCRIPT LOAD again
5.  Cache is flushed again
6.  We fail again.

If we just attempt the EVALSHA and then failover to calling EVAL we eliminate any possible race condition as EVAL will contain the code we want Redis to execute _and_ will cause Redis to cache the script.  This will mean that the very first EVALSHA to a specific Redis server will need to failover, but all subsequent calls will succeed using EVALSHA.

It also means we don't need to store the two new REDIS_G globals (is the script uploaded and what the sha1 hash is).  The sha1 digest is deterministic.  We know what it is from the LUA itself.

Thanks again for the pull request.  Lots of people are going to find it useful.

SkydiveMarius commented:
@michael-grunder Thank you for your review. 
Your suggestions regarding LUA sounds good and I will try to implement it.

> I've pulled the branch in locally and have made a few small changes having mostly to do with formatting but also a couple things to make the code more "C like" here and there.

Could you please create a branch containing this changes, so that I can continue based on the changes you made?
As an alternative I granted push permissions to my fork, so you might push them directly to my PR branch.

michael-grunder commented:
@SkydiveMarius I'm going to try to get your branch with the small modifications up in the next day or two at which point everyone that's interested will hopefully help us test it!

michael-grunder commented:
@SkydiveMarius Like I said, I pulled your branch down locally and made a few changes and it now lives [here](https://github.com/phpredis/phpredis/tree/session-locking).

A lot of the changes were just reorganizing how the new functions were declared, etc.  The only big logical change I made was to use `EVALSHA` with an `EVAL` fallback rather than trying to use `SCRIPT LOAD`.

I was able to do a bit of testing with this locking functionality using various different configuration parameters and it looks pretty good!   This really is a nice improvement that people have wanted for along time!

SInce this is a reasonably big functional change I think it best to let the code live in a separate branch for some time so we can get more people to test it and provide feedback.

Closing this pull request as your changes are incorporated into the new branch.
